### PR TITLE
RELEASES: add formal release policy for runc

### PR DESCRIPTION
**Pull request by cyphar**
Historically, our release cadence and support policy has been quite
ad-hoc, which has caused some strife and is not really becoming of a
project as widely used as ours. So, in an attempt to make our releases
more regular and to provide more guidance on our backport policy, add a
document outlining our policy.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

### CODE REVIEW

**Comment by cyphar** [comment]
As this is more of a governance-related discussion, it would be nice to hear the opinions of (and once we've agreed on a potential policy, get approvals from) from two-thirds of maintainers (4):

 - [ ] Mrunal Patel <mpatel@redhat.com> (@mrunalp)
 - [ ] Aleksa Sarai <cyphar@cyphar.com> (@cyphar)
 - [ ] Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp> (@AkihiroSuda)
 - [ ] Kir Kolyshkin <kolyshkin@gmail.com> (@kolyshkin)
 - [ ] Sebastiaan van Stijn <github@gone.nl> (@thaJeztah)
 - [ ] Li Fu Bang <lifubang@acmcoder.com> (@lifubang)
 - [ ] Rodrigo Campos <rodrigoca@microsoft.com> (@rata)

File: RELEASES.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   been assessed as having CVSS score 7 or higher). At this point, users still
+   using `latest-2` would be strongly encouraged to upgrade to either `latest`
+   or `latest-1`.
+
+ * Any older releases will no longer receive any updates, and users are
+   encouraged to upgrade in the strongest possible terms, as they will not
+   receive any security fixes regardless of severity or impact.
+
+This policy only applies to minor releases of runc with major version `1`. In
+the unlikely event of a runc `2.0` release, this document will be updated to
```
Comment (AkihiroSuda) [inline RELEASES.md:91]: It is likely that we will release v2.0 when we drop cgroup v1 in 2030-ish

File: RELEASES.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+compatibility policy only applies to the runc binary. We will make a
+best-effort attempt to reduce the impact to users that make direct use of
+`github.com/opencontainers/runc/libcontainer` as a Go library, but we do not
+formally guarantee that API compatibility will be preserved.
+
+[semver]: https://semver.org/spec/v2.0.0.html
+
+### Release Cadence ###
+
+runc follows a 6-month minor version release schedule, with the aim of releases
```
Comment (AkihiroSuda) [inline RELEASES.md:21]: This should be probably marked as a draft proposal until we actually release v1.3, to verify the feasibility of the plan.

File: RELEASES.md (hunk 1 above)
Comment (cyphar) [reply RELEASES.md:91]: I'll drop the word "unlikely" but 2030 is still a while away. :sweat_smile:

File: RELEASES.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+most widely used Linux container runtime, our release schedule has been very
+ad-hoc and has resulted in very long periods of time between minor releases,
+causing issues for downstreams that wanted particular features.
+
+### Semantic Versioning ###
+
+runc uses [Semantic Versioning][semver] for releases. However, our
+compatibility policy only applies to the runc binary. We will make a
+best-effort attempt to reduce the impact to users that make direct use of
+`github.com/opencontainers/runc/libcontainer` as a Go library, but we do not
```
Comment (kolyshkin) [inline RELEASES.md:14]: nit: I'd rephrase this, as "a Go library" is probably not a universally understood term.

Something similar to this maybe:
```suggestion
Go packages prefixed with `github.com/opencontainers/runc`, but we do not
```

**Review by kolyshkin — APPROVED** [review]
LGTM

File: RELEASES.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+However, the following release will still go according to schedule (except in
+the exceptionally unlikely scenario where the delay is 4-6 months long, in
+which case the next release is moved forward to when the subsequent release
+would have been).
+
+Here is a hypothetical release timeline to see how this works in practice:
+
+| Date       | Release      | Notes |
+| ---------- | ------------ | ----- |
+| 200X-02-28 | `1.3.0-rc.1` | `release-1.3` branch created, feature freeze. |
```
Comment (lifubang) [inline RELEASES.md:52]: According to this release policy, the version of `1.3.0-rc.1` will be released in `2025-02-28`. If I understand right, all of tasks in [1.3.0 milestone](https://github.com/opencontainers/runc/milestone/14) should be resolved before we release `1.3.0-rc.1`. There are only about 2.5 months left, I think it's hard to complete it. Maybe we should reorganize the task list for this milestone.

Comment (cyphar) [reply RELEASES.md:52]: If we agree on doing time-based releases then we should move any outstanding items to later release milestones if they don't make the cut. (Also, quite a few of those were added to 1.3 to indicate they were post-1.2 not that they needed to make it into 1.3.)

Comment (kolyshkin) [reply RELEASES.md:52]: Yes, and we did the same for 1.2 -- moved some PRs/issues to a later release.

With time-based releases, instead of trying to ensure everything that's marked with the next milestone is going in (and delaying a release by a couple of years in the process), we fix/merge what we can, and change the milestone to the next one for the rest. This way we're not perfect (well,we never are) but at least we're predictable wrt when a new release is made.

**Review by lifubang — APPROVED** [review]

**Comment by cyphar** [comment]
We've hit the required LGTM threshold, but we can hold off a bit on merging for a bit, to see if @mrunalp @thaJeztah or @rata had any comments.

**Review by rata — APPROVED** [review]
LGTM, thanks!

Sorry, I've been out for a long time and I'll still be away until mid January or so.

File: RELEASES.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+minor release is published. In the following text, `latest` refers to the
+latest minor (non-release-candidate) runc release published; `latest-1` is the
+previous minor release branch; and `latest-2` is the minor release branch
+before `latest-1`. For example, if `latest` is `1.4.0` then `latest-1` is
+`1.3.z` and `latest-2` is `1.2.z`.
+
+ * Once `latest` is released, new features will no longer be merged into
+   `latest` and only bug and security fixes will be backported, though we will
+   be fairly liberal with what kinds of bugs will considered candidates for
+   backporting.
```
[the remark this answers is not in the document]
Comment (cyphar) [reply RELEASES.md:89]: It is, but given that we haven't always followed this in the past, it might be nice to make this explicit.

File: RELEASES.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+| ---------- | ------------ | ----- |
+| 200X-02-28 | `1.3.0-rc.1` | `release-1.3` branch created, feature freeze. |
+| 200X-03-12 | `1.3.0-rc.2` | |
+| 200X-03-25 | `1.3.0-rc.3` | |
+| 200X-04-30 | `1.3.0`      | `1.3` release published. |
+| 200X-05-10 | `1.3.1`      | |
+| 200X-06-21 | `1.3.2`      | |
+| 200X-06-25 | `1.3.3`      | |
+| 200X-07-02 | `1.3.4`      | |
+| 200X-08-28 | `1.4.0-rc.1` | `release-1.4` branch created, feature freeze. |
```
[the remark this answers is not in the document]
Comment (cyphar) [reply RELEASES.md:60]: I would personally prefer that we end up with a minor release without any new features rather than a delayed minor release because we didn't have a strict enough deadline.

IMHO even if a minor release doesn't have many (or any) new features that's fine. SemVer says that minor bumps MUST be done for features, but it doesn't say they MUST NOT be done in absence of features. And while it is a minor thing, doing minor releases means that we shorten the lifespan of release branches, which makes backports less painful.

IMHO we've already tried to do the "we'll release once we have a new feature that deserves a release" approach and that doesn't seem to work for us (to be fair, we weren't strict enough about deadlines but I think a deadline decided that way is going to end up being somewhat loose). Having an automated deadline decided long in advance without needing to think about IMHO would make it easier for us and our users to plan around (not to mention it will make revert decisions _much_ easier). It also means we only need to agree on this timeline once for the life of the project, rather than getting agreement for each minor release.

**Review by mrunalp — APPROVED** [review]

**Comment by mrunalp** [comment]
Thanks for driving this. LGTM!

**Comment by kolyshkin** [comment]
With the approval of the majority of maintainers, let's merge this now. If there are any followups, those can be raised in separate subsequent PRs.
